File: web/src/import/diagnostics.ts
// The report a listener can choose to send when an archive fails closed:
// file names, byte sizes, row counts, and top-level key names. Never a value.

import { PARSER_VERSION, loadRecords, planArchive } from './spotify-parser'
import type { ExportArchive } from './zip-reader'

export type DiagnosticsFile = {
  path: string
  bytes: number
  /** Row count for a file the package reads and decodes; null otherwise. */
  rows: number | null
  /** Top-level keys of the first array element (history) or of the object (library, playlist); null when not read or broken. */
  headers: string[] | null
}

export type ExportDiagnostics = {
  source: 'spotify_export'
  parserVersion: string
  files: DiagnosticsFile[]
}

export type DiagnoseOptions = {
  signal?: AbortSignal
}

export async function diagnoseExport(archive: ExportArchive, options: DiagnoseOptions = {}): Promise<ExportDiagnostics> {
  const plan = await planArchive(archive)
  const files: DiagnosticsFile[] = []
  for (const file of plan.files) {
    if (!file.read || file.kind === null) {
      files.push({ path: file.path, bytes: file.bytes, rows: null, headers: null })
      continue
    }
    options.signal?.throwIfAborted()
    let rows: number | null = null
    let headers: string[] | null = null
    try {
      const loaded = loadRecords(file.kind, await archive.readText(file.path, { signal: options.signal }))
      rows = loaded.rows
      headers = loaded.headers
    } catch (error) {
      options.signal?.throwIfAborted()
      void error
    }
    files.push({ path: file.path, bytes: file.bytes, rows, headers })
  }
  return { source: 'spotify_export', parserVersion: PARSER_VERSION, files }
}

File: client/lib/import/diagnostics.dart
import 'dart:io';
import 'dart:isolate';

import 'snapshot.dart';
import 'spotify_parser.dart';
import 'zip_reader.dart';

/// Recorded with a failed import so a report can be read without the archive.
const String spotifyParserVersion = 'client-spotify-1';

/// Names, byte sizes, row counts, and top-level keys; never a value.
///
/// Rows and headers are filled only for the files the detected package reads
/// (the same set the inventory lists as `read`); every other entry, allow-listed
/// or not, is reported by path and size alone. Headers are the keys of the first
/// array element of a history file, or of the library / playlist object.
Future<ExportDiagnostics> diagnoseExport(ExportArchive archive) async {
  final files = await listExportFiles(archive);
  final package = detectPackage(files);
  final readKinds = readKindsFor(package);
  final report = <DiagnosticsFile>[];
  for (final file in files) {
    int? rows;
    List<String>? headers;
    if (readKinds.contains(file.kind)) {
      final loaded = await loadExportFile(archive, file);
      if (loaded != null) {
        rows = loaded.rows;
        headers = loaded.headers;
      }
    }
    report.add(DiagnosticsFile(path: file.path, bytes: file.bytes, rows: rows, headers: headers));
  }
  return ExportDiagnostics(
    source: package == null ? 'unknown' : 'spotify_export',
    files: report,
    parserVersion: spotifyParserVersion,
  );
}

/// [diagnoseExport] over the archive at [path], off the main isolate (it
/// decodes every file the parser would read, which for a long history is
/// seconds of work). A file that is not a ZIP reports `unknown` with no
/// files.
Future<ExportDiagnostics> diagnoseExportFile(String path) => Isolate.run(() => _diagnoseFile(path));

Future<ExportDiagnostics> _diagnoseFile(String path) async {
  final ZipExportArchive archive;
  try {
    archive = ZipExportArchive.open(File(path));
  } on ArchiveFormatException {
    return const ExportDiagnostics(source: 'unknown', files: [], parserVersion: spotifyParserVersion);
  }
  try {
    return await diagnoseExport(archive);
  } finally {
    await archive.close();
  }
}
